tools/redir-gen/main.go

```diff
@@ -0,0 +1,184 @@
+	allRepos := riSlice{}
+	client := github.NewClient(nil)
+	for _, org := range orgs {
+		opts := &github.RepositoryListByOrgOptions{
+			ListOptions: github.ListOptions{PerPage: 50},
+		}
+		for {
+			repos, resp, err := client.Repositories.ListByOrg(ctx, org, opts)
+			if err != nil {
+				return nil, err
+			}
+			for _, r := range repos {
+				if !allowedRepoRe.MatchString(*r.Name) {
+					log.Printf("Ignoring repo %s, matched by ignore %q", *r.Name, allowedRepoRe)
+					continue
+				}
+				if *r.Archived {
+					log.Print("Ignoring archived repo: ", *r.Name)
+					continue
+				}
+				allRepos = append(allRepos, repoInfo{org, *r.Name, *r.DefaultBranch})
+			}
+			if resp.NextPage == 0 {
+				break
+			}
+		}
+	}
+	sort.Sort(allRepos)
+	return allRepos, nil
+}
+
+const (
+	goHTML = `<html><head>
+    <meta name="go-import" content="knative.dev/{{.Repo}} git https://github.com/{{.Org}}/{{.Repo}}">
+    <meta name="go-source" content="knative.dev/{{.Repo}}     https://github.com/{{.Org}}/{{.Repo}} https://github.com/{{.Org}}/{{.Repo}}/tree/{{.DefaultBranch}}{/dir} https://github.com/{{.Org}}/{{.Repo}}/blob/{{.DefaultBranch}}{/dir}/{file}#L{line}">
+</head></html>
+`
+	redirText = `/{{.Repo}}/* go-get=1 /golang/{{.Repo}}.html 200
+`
+
+	autogenPrefix = `
+# This file is AUTO-GENERATED
+#
+# DO NOT EDIT!
+#
+# To regenerate, run:
+`
+)
+
+func appendRedirs(ris []repoInfo) error {
+	redirFilename := "golang/_redirects"
+	redirFile, err := os.OpenFile(redirFilename, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return fmt.Errorf("unable to open %q: %w", redirFilename, err)
+	}
+	defer redirFile.Close()
+
+	redirs, err := ioutil.ReadAll(redirFile)
+	if err != nil {
+		return fmt.Errorf("unable to read %q: %w", redirFilename, err)
+	}
+
+	seekTo := int64(strings.Index(string(redirs), autogenPrefix))
+	if seekTo == -1 {
+		seekTo = int64(len(redirs))
+	}
+	if _, err := redirFile.Seek(seekTo, 0); err != nil {
+		return fmt.Errorf("unable to seek in %q: %w", redirFilename, err)
+	}
+	if redirFile.Truncate(seekTo) != nil {
+		return fmt.Errorf("unable to truncate %q: %w", redirFilename, err)
+	}
+
+	if _, err := redirFile.WriteString(autogenPrefix); err != nil {
+		return fmt.Errorf("unable to write to %q: %w", redirFilename, err)
+	}
+	if _, err := redirFile.WriteString(fmt.Sprintf("#   %s\n", strings.Join(os.Args, " "))); err != nil {
+		return fmt.Errorf("unable to write to %q: %w", redirFilename, err)
+	}
+	for _, ri := range ris {
+		if redirTemplate.Execute(redirFile, ri) != nil {
+			return fmt.Errorf("unable to write %s to %q: %w", ri.Repo, redirFilename, err)
+		}
+	}
+	return nil
+}
+
+var fileTemplate = template.Must(template.New("gohtml").Parse(goHTML))
+var redirTemplate = template.Must(template.New("redir").Parse(redirText))
+
+// createGoGetFile creates a static HTML file providing a knative.dev mapping
+// for the specified org and repo.
+func createGoGetFile(ri repoInfo) error {
+	filename := fmt.Sprintf("golang/%s.html", ri.Repo)
```

Comment:
```suggestion
	dirPath := "golang"
	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
		if err = os.MkdirAll(dirPath, 0777); err != nil {
			return fmt.Errorf("error creating directory: %w", err)
		}
	}
	filename := fmt.Sprintf("%s/%s.html", dirPath, ri.Repo)
```

Reply:
you can address this in another PR if you want to address it

Reply:
This is a straight copy/paste of the existing redir-gen in website repo; while we _could_ tidy it up here/improve it, I'd rather do that separately. As I say above, the directory must exist, or this is being run from the wrong location and it's better to error in that case.

(FWIW if we did want to create the directory there's no need to do the os.Stat since os.MkdirAll does not error if the directory already exists. It would also probably be better to use `filepath.Join` to create the filename)

Reply:
I felt that whatever I pointed out how to create the directory was going to be wrong.
I looked up an example from the knative source code first :-)
https://github.com/knative/serving/blob/561d9d0c6a04256fcd56fd304107fbe3b2e77dff/vendor/knative.dev/pkg/test/helpers/dir.go#L37-L38

Reply:
Heh, I might PR a tidy up of that tomorrow then. Stat + mkdir makes sense, or just MkdirAll on its own, but stat + mkdirall is redundant - mkdirall already does a stat. I got worried I was missing something for a sec but StackOverflow agrees 😂 https://stackoverflow.com/questions/35661967/checking-with-os-stat-vs-just-os-mkdirall.